#! /bin/sh -e
##
## All lines beginning with `## DP:' are a description of the patch.
##
## DP: Fix security issue related to bad handling of cookies for session authentication

[ -f debian/patches/00patch-opts ] && . debian/patches/00patch-opts
patch_opts="${patch_opts:--f --no-backup-if-mismatch}"

if [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
       -patch)   ( cd BUILD/sql-ledger && patch $patch_opts -p3 ) < $0 ;;
       -unpatch) ( cd BUILD/sql-ledger && patch $patch_opts -p3 -R ) < $0 ;;
        *)
                echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
                exit 1;;
esac

exit 0
@DPATCH@
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/am.pl /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/am.pl
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/am.pl	2006-04-14 01:08:53.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/am.pl	2006-09-09 17:10:06.000000000 +0200
@@ -8,7 +8,7 @@
 #   Email: dsimader@sql-ledger.org
 #     Web: http://www.sql-ledger.org
 #
-#  Contributors:
+#  Contributors:	Tony Fraser <tony@sybaspace.com>
 #
 # This program is free software; you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
@@ -72,6 +72,7 @@
   $msg1 = $locale->text('You are logged out!');
   $msg2 = $locale->text('Login');
   $form->redirect("$msg1 <p><a href=login.pl target=_top>$msg2</a>");
+  exit;
 }
 
 # locale messages
@@ -144,8 +145,19 @@
 	  $form->error($locale->text('Access Denied!'));
 	}
 	exit;
+      } else {
+        # password checked out, create session
+	if ($ENV{HTTP_USER_AGENT}) {
+	  # create new session
+	  use SL::User;
+	  $user = new User $memberfile, $form->{login};
+	  $user->{password} = $form->{password};
+	  $user->create_config("$userspath/$form->{login}.conf");
+	  $form->{sessioncookie} = $user->{sessioncookie};
+	}
       }
     } else {
+      
       if ($ENV{HTTP_USER_AGENT}) {
 	$ENV{HTTP_COOKIE} =~ s/;\s*/;/g;
 	@cookies = split /;/, $ENV{HTTP_COOKIE};
@@ -155,9 +167,40 @@
 	}
 	
 	if ($form->{action} ne 'display') {
-	  if ((! $cookie{"SQL-Ledger-$form->{login}"}) || $cookie{"SQL-Ledger-$form->{login}"} ne $form->{sessionid}) {
-	    &getpassword(1);
-	    exit;
+	  if ($cookie{"SL-$form->{login}"}) {
+	    
+	    $form->{sessioncookie} = $cookie{"SL-$form->{login}"};
+	    
+	    $s = "";
+	    %ndx = ();
+            # take cookie apart
+	    $l = length $form->{sessioncookie};
+	    
+	    for $i (0 .. $l - 1) {
+	      $j = substr($myconfig{sessionkey}, $i * 2, 2);
+	      $ndx{$j} = substr($cookie{"SL-$form->{login}"}, $i, 1);
+	    }
+
+	    for (sort keys %ndx) {
+	      $s .= $ndx{$_};
+	    }
+
+            $l = length $form->{login};
+	    $login = substr($s, 0, $l);
+	    $time = substr($s, -10);
+	    $password = substr($s, $l, (length $s) - ($l + 10));
+
+	    # validate cookie
+	    if ((time > $time) || ($login ne $form->{login}) || ($myconfig{password} ne crypt $password, substr($form->{login}, 0, 2))) {
+	      &getpassword(1);
+	      exit;
+	    }
+
+	  } else {
+
+	    &getpassword(1); 
+	    exit; 
+
 	  }
 	}
       } else {
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/admin.pl /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/admin.pl
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/admin.pl	2006-06-13 17:35:52.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/admin.pl	2006-09-10 20:00:18.000000000 +0200
@@ -42,6 +42,7 @@
 $form->{stylesheet} = "sql-ledger.css";
 $form->{favicon} = "favicon.ico";
 $form->{timeout} = 600;
+$form->{"root login"} = 1;
 
 require "$form->{path}/pw.pl";
 
@@ -86,7 +87,7 @@
 sub adminlogin {
 
   $form->{title} = qq|SQL-Ledger $form->{version} |.$locale->text('Administration');
-  
+
   $form->header;
   
     print qq|
@@ -130,8 +131,47 @@
 }
 
 
+sub create_config {
+  
+  $form->{sessionkey} = "";
+  $form->{sessioncookie} = "";
+  
+  if ($form->{password}) {
+    my $t = time + $form->{timeout};
+    srand( time() ^ ($$ + ($$ << 15)) );
+    $key = "root login$form->{password}$t";
+    
+    my $i = 0;
+    my $l = length $key;
+    my $j = $l;
+    my %ndx = ();
+    my $pos;
+    
+    while ($j > 0) {
+      $pos = int rand($l);
+      next if $ndx{$pos};
+      $ndx{$pos} = 1;
+      $form->{sessioncookie} .= substr($key, $pos, 1);
+      $form->{sessionkey} .= substr("0$pos", -2);
+      $j--;
+    }
+  }
+
+  open(CONF, ">$userspath/root login.conf") or $form->error("root login.conf : $!");
+  print CONF qq|# configuration file for root login
+
+\%rootconfig = (
+sessionkey => '$form->{sessionkey}'
+);\n\n|;
+  
+  close CONF;
+
+}
+
+
 sub login {
 
+  &create_config;
   &list_users;
 
 }
@@ -139,7 +179,7 @@
 
 sub logout {
 
-  $form->{callback} = "$form->{script}?path=$form->{path}&endsession=1";
+  $form->{callback} = "$form->{script}?path=$form->{path}";
   $form->redirect($locale->text('You are logged out'));
 
 }
@@ -254,7 +294,6 @@
 
   $form->{title} = "SQL-Ledger ".$locale->text('Accounting')." ".$locale->text('Administration');
 
-  $form->{login} = "root login";
   $form->header;
 
   print qq|
@@ -280,7 +319,7 @@
 |;
 
 foreach $key (sort keys %member) {
-  $href = "$script?action=edit&login=$key&path=$form->{path}&sessionid=$form->{sessionid}";
+  $href = "$script?action=edit&login=$key&path=$form->{path}";
   $href =~ s/ /%20/g;
   
   $member{$key}{templates} =~ s/^$templates\///;
@@ -447,10 +486,7 @@
 
   }
   
-  $user = $form->{login};
-  $form->{login} = "root login";
   $form->header;
-  $form->{login} = $user;
  
   print qq|
 <body class=admin>
@@ -740,6 +776,7 @@
   # no driver checked
   $form->error($locale->text('Database Driver not checked!')) unless $form->{dbdriver};
 
+ 
   # no spaces allowed in login name
   $form->{login} =~ s/ //g;
 
@@ -805,13 +842,14 @@
     $form->isblank("dbname", $locale->text('Dataset missing!'));
     $form->isblank("dbuser", $locale->text('Database User missing!'));
   }
-    
+  
   foreach $item (keys %{$form}) {
     $myconfig->{$item} = $form->{$item};
   }
 
-  $myconfig->{password} = $form->{old_password};
-  $myconfig->{password} = $form->{new_password} if $form->{new_password} ne $form->{old_password};
+  $myconfig->{encrypted} = $form->{new_password} eq $form->{old_password};
+  $myconfig->{password} = $form->{new_password};
+
   $myconfig->{timeout} = $form->{newtimeout};
 
   delete $myconfig->{stylesheet};
@@ -820,6 +858,7 @@
   }
   
   $myconfig->{packpw} = 1;
+  delete $myconfig->{"root login"};
   
   $myconfig->save_member($memberfile, $userspath);
 
@@ -973,7 +1012,6 @@
 
   $form->{title} = qq|SQL-Ledger |.$locale->text('Accounting')." ".$locale->text('Administration')." / ".$locale->text('Change Admin Password');
 
-  $form->{login} = "root login";
   $form->header;
 
   print qq|
@@ -1028,9 +1066,13 @@
   $root->{password} = $form->{new_password};
   
   $root->{'root login'} = 1;
+  $root->{login} = "root login";
   $root->save_member($memberfile);
 
-  $form->{callback} = "$form->{script}?action=list_users&path=$form->{path}&sessionid=$form->{sessionid}";
+  $form->{password} = $form->{new_password};
+  &create_config;
+  
+  $form->{callback} = "$form->{script}?action=list_users&path=$form->{path}&sessionid=$form->{sessionid}&password=$form->{password}";
 
   $form->redirect($locale->text('Password changed!'));
 
@@ -1041,22 +1083,62 @@
 
   $root = new User "$memberfile", "root login";
 
+  $rootname = "root login";
+  eval { require "$userspath/${rootname}.conf"; };
+  
   if ($root->{password}) {
-      
+   
     if ($form->{password}) {
       $form->{callback} .= "&password=$form->{password}" if $form->{callback};
-      $form->{sessionid} = time;
       if ($root->{password} ne crypt $form->{password}, 'ro') {
 	&getpassword;
 	exit;
       }
+
+      # create config
+      &create_config;
+      
     } else {
+
       if ($ENV{HTTP_USER_AGENT}) {
 	$ENV{HTTP_COOKIE} =~ s/;\s*/;/g;
-	%cookie = split /[=;]/, $ENV{HTTP_COOKIE};
-	$cookie = ($form->{path} eq 'bin/lynx') ? $cookie{login} : $cookie{"SQL-Ledger-root login"};
-	if (! $cookie || $cookie ne $form->{sessionid}) {
-	  &getpassword;
+	@cookies = split /;/, $ENV{HTTP_COOKIE};
+	%cookie = ();
+	foreach (@cookies) {
+	  ($name,$value) = split /=/, $_, 2;
+	  $cookie{$name} = $value;
+	}
+ 
+	$cookie = ($form->{path} eq 'bin/lynx') ? $cookie{login} : $cookie{"SL-root login"};
+
+	if ($cookie) {
+	  $form->{sessioncookie} = $cookie;
+
+	  $s = ""; 
+	  %ndx = (); 
+	  $l = length $form->{sessioncookie};
+
+	  for $i (0 .. $l - 1) {
+	    $j = substr($rootconfig{sessionkey}, $i * 2, 2);
+	    $ndx{$j} = substr($cookie, $i, 1);
+	  }
+
+	  for (sort keys %ndx) {
+	    $s .= $ndx{$_}; 
+	  }
+
+	  $l = length 'root login';
+	  $login = substr($s, 0, $l);
+	  $time = substr($s, -10);
+	  $password = substr($s, $l, (length $s) - ($l + 10));
+
+	  if ((time > $time) || ($login ne 'root login') || ($root->{password} ne crypt $password, 'ro')) {
+	    &getpassword; 
+	    exit; 
+	  }
+	  
+	} else {
+	  &getpassword; 
 	  exit;
 	}
       }
@@ -1120,7 +1202,6 @@
 
  $form->{title} = "SQL-Ledger ".$locale->text('Accounting')." / ".$locale->text('Database Administration');
   
-  $form->{login} = "root login";
   $form->header;
 
   print qq|
@@ -1204,7 +1285,6 @@
 
   $form->{title} = "SQL-Ledger ".$locale->text('Accounting')." ".$locale->text('Database Administration')." / ".$locale->text('Update Dataset');
   
-  $form->{login} = "root login";
   $form->header;
 
   print qq|
@@ -1343,7 +1423,6 @@
   
   $form->{title} = "SQL-Ledger ".$locale->text('Accounting')." ".$locale->text('Database Administration')." / ".$locale->text('Create Dataset');
   
-  $form->{login} = "root login";
   $form->header;
 
   print qq|
@@ -1447,7 +1526,6 @@
   
   $form->{title} = "SQL-Ledger ".$locale->text('Accounting')." ".$locale->text('Database Administration')." / ".$locale->text('Create Dataset');
 
-  $form->{login} = "root login";
   $form->header;
 
   print qq|
@@ -1491,7 +1569,6 @@
 
   $form->{title} = "SQL-Ledger ".$locale->text('Accounting')." ".$locale->text('Database Administration')." / ".$locale->text('Delete Dataset');
 
-  $form->{login} = "root login";
   $form->header;
 
   print qq|
@@ -1555,7 +1632,6 @@
 
   $form->{title} = "SQL-Ledger ".$locale->text('Accounting')." ".$locale->text('Database Administration')." / ".$locale->text('Delete Dataset');
 
-  $form->{login} = "root login";
   $form->header;
 
   print qq|
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/am.pl /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/am.pl
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/am.pl	2006-07-01 19:35:04.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/am.pl	2006-09-05 17:44:05.000000000 +0200
@@ -2068,7 +2068,6 @@
 
 <form method=post action=$form->{script}>
 
-<input type=hidden name=old_password value=$myconfig{password}>
 <input type=hidden name=type value=preferences>
 <input type=hidden name=role value=$myconfig{role}>
 
@@ -2114,7 +2113,7 @@
 	    <table>
 	      <tr>
 		<th align=right>|.$locale->text('Password').qq|</th>
-		<td><input type=password name=new_password size=10 value=$myconfig{password}></td>
+		<td><input type=password name=new_password size=10 value=*></td>
 	      </tr>
 	      <tr>
 		<th align=right>|.$locale->text('Confirm').qq|</th>
@@ -2206,9 +2205,9 @@
 
   $form->{stylesheet} = $form->{usestylesheet};
 
-  if ($form->{new_password} ne $form->{old_password}) {
-    $form->error($locale->text('Password does not match!')) if $form->{new_password} ne $form->{confirm_password};
-  }
+  $form->error($locale->text('Password does not match!')) if $form->{new_password} ne $form->{confirm_password};
+
+  $form->{password} = $form->{new_password};
 
   if (AM->save_preferences(\%myconfig, \%$form, $memberfile, $userspath)) {
     $form->redirect($locale->text('Preferences saved!'));
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/login.pl /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/login.pl
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/login.pl	2006-01-19 23:55:42.000000000 +0100
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/login.pl	2006-09-10 03:12:28.000000000 +0200
@@ -5,6 +5,7 @@
 #  Author: DWS Systems Inc.
 #     Web: http://www.sql-ledger.org
 #
+#  Contributors:
 #
 # This program is free software; you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
@@ -67,7 +68,6 @@
   $form->{stylesheet} = "sql-ledger.css";
   $form->{favicon} = "sql-ledger.ico";
 
-  $form->{endsession} = 1;
   $form->header(1);
 
   if ($form->{login}) {
@@ -158,6 +158,11 @@
     $form->{stylesheet} = "sql-ledger.css";
   }
 
+  if (-f sql-ledger.ico) {
+    $form->{favicon} = "sql-ledger.ico";
+  }
+  
+  delete $self->{sessioncookie};
   $form->header(1);
 
   print qq|
@@ -300,7 +305,7 @@
 
       print $locale->text('done');
 
-      print "<p><a href=menu.pl?login=$form->{login}&sessionid=$form->{sessionid}&path=$form->{path}&action=display&main=company_logo&js=$form->{js}>".$locale->text('Continue')."</a>";
+      print "<p><a href=menu.pl?login=$form->{login}&path=$form->{path}&action=display&main=company_logo&js=$form->{js}>".$locale->text('Continue')."</a>";
 
       exit;
     }
@@ -308,9 +313,13 @@
     $form->error($err[$errno]);
   }
 
+  $user->create_config("$userspath/$form->{login}.conf");
+  $form->{timeout} = $user->{timeout};
+  $form->{sessioncookie} = $user->{sessioncookie};
+
   # made it this far, setup callback for the menu
   $form->{callback} = "menu.pl?action=display";
-  for (qw(login path js)) { $form->{callback} .= "&$_=$form->{$_}" }
+  for (qw(login path js sessioncookie)) { $form->{callback} .= "&$_=$form->{$_}" }
   
   # check for recurring transactions
   if ($user->{acs} !~ /Recurring Transactions/) {
@@ -340,10 +349,7 @@
 sub logout {
 
   $form->{callback} = "$form->{script}?path=$form->{path}&login=$form->{login}";
-
-  $form->{endsession} = 1;
-  
-  $form->redirect;
+  $form->redirect($locale->text('You are logged out'));
 
 }
 
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/menu.pl /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/menu.pl
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/menu.pl	2006-03-25 02:58:33.000000000 +0100
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/menu.pl	2006-09-10 03:08:45.000000000 +0200
@@ -39,14 +39,14 @@
   $menuwidth = ($ENV{HTTP_USER_AGENT} =~ /links/i) ? "240" : "155";
   $menuwidth = $myconfig{menuwidth} if $myconfig{menuwidth};
 
-  $form->header(!$form->{duplicate});
+  $form->header;
 
   print qq|
 
 <FRAMESET COLS="$menuwidth,*" BORDER="1">
 
-  <FRAME NAME="acc_menu" SRC="$form->{script}?login=$form->{login}&sessionid=$form->{sessionid}&action=acc_menu&path=$form->{path}&js=$form->{js}">
-  <FRAME NAME="main_window" SRC="am.pl?login=$form->{login}&sessionid=$form->{sessionid}&action=$form->{main}&path=$form->{path}">
+  <FRAME NAME="acc_menu" SRC="$form->{script}?login=$form->{login}&action=acc_menu&path=$form->{path}&js=$form->{js}">
+  <FRAME NAME="main_window" SRC="am.pl?login=$form->{login}&action=$form->{main}&path=$form->{path}">
 
 </FRAMESET>
 
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/pw.pl /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/pw.pl
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/bin/mozilla/pw.pl	2005-07-12 06:42:43.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/bin/mozilla/pw.pl	2006-09-08 05:45:13.000000000 +0200
@@ -28,8 +28,30 @@
 sub getpassword {
   my ($s) = @_;
 
-  $form->{endsession} = 1;
-  $form->header;
+  my $login = ($form->{"root login"}) ? "root login" : $form->{login};
+  
+  my @d = split / +/, scalar gmtime(time);
+  my $today = "$d[0], $d[2]-$d[1]-$d[4] $d[3] GMT";
+  
+  if ($form->{stylesheet} && (-f "css/$form->{stylesheet}")) {
+    $stylesheet = qq|<LINK REL="stylesheet" HREF="css/$form->{stylesheet}" TYPE="text/css" TITLE="SQL-Ledger stylesheet">
+|;
+  }
+  
+  if ($form->{charset}) {
+    $charset = qq|<META HTTP-EQUIV="Content-Type" CONTENT="text/plain; charset=$form->{charset}">
+|;
+  }
+
+  print qq|Set-Cookie: SL-$login=; expires=$today; path=/;
+Content-Type: text/html
+
+<head>
+  <title>$form->{titlebar}</title>
+  $stylesheet
+  $charset
+</head>
+|;
 
   $sessionexpired = qq|<b><font color=red><blink>|.$locale->text('Session expired!').qq|</blink></font></b><p>| if $s;
   
@@ -58,7 +80,7 @@
 
 |;
 
-  for (qw(script endsession password)) { delete $form->{$_} }
+  for (qw(script password header)) { delete $form->{$_} }
   $form->hide_form;
   
   print qq|
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/AM.pm /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/AM.pm
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/AM.pm	2006-08-25 19:30:52.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/AM.pm	2006-09-10 00:37:58.000000000 +0200
@@ -1091,10 +1091,9 @@
   foreach my $item (keys %$form) {
     $myconfig->{$item} = $form->{$item};
   }
-  
-  $myconfig->{password} = $form->{new_password} if ($form->{old_password} ne $form->{new_password});
 
   $myconfig->save_member($memberfile, $userspath);
+  $form->{sessioncookie} = $myconfig->{sessioncookie};
 
   1;
 
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/Form.pm /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/Form.pm
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/Form.pm	2006-08-25 19:34:04.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/Form.pm	2006-09-09 19:54:59.000000000 +0200
@@ -235,7 +235,7 @@
   
 
 sub header {
-  my ($self, $init) = @_;
+  my ($self, $endsession) = @_;
 
   return if $self->{header};
 
@@ -261,12 +261,13 @@
 
     $self->{titlebar} = ($self->{title}) ? "$self->{title} - $self->{titlebar}" : $self->{titlebar};
 
-    $self->set_cookie($init);
+    $self->set_cookie($endsession);
 
     print qq|Content-Type: text/html
 
 <head>
   <title>$self->{titlebar}</title>
+  <META NAME="robots" CONTENT="noindex,nofollow" />
   $favicon
   $stylesheet
   $charset
@@ -282,25 +283,29 @@
 
 
 sub set_cookie {
-  my ($self, $init) = @_;
+  my ($self, $endsession) = @_;
 
-  $self->{timeout} = ($self->{timeout} > 0) ? $self->{timeout} : 3600;
-  my $t = ($self->{endsession}) ? time : time + $self->{timeout};
+  $self->{timeout} ||= 3600;
+  my $t = ($endsession) ? time : time + $self->{timeout};
+  my $login = ($self->{"root login"}) ? "root login" : $self->{login};
   
   if ($ENV{HTTP_USER_AGENT}) {
-
     my @d = split / +/, scalar gmtime($t);
     my $today = "$d[0], $d[2]-$d[1]-$d[4] $d[3] GMT";
 
-    if ($init) {
-      $self->{sessionid} = time;
+    if ($login) {
+      if ($self->{sessioncookie}) {
+	print qq|Set-Cookie: SL-${login}=$self->{sessioncookie}; expires=$today; path=/;\n|;
+      } else {
+	print qq|Set-Cookie: SL-${login}=; expires=$today; path=/;\n|;
+      }
     }
-    print qq|Set-Cookie: SQL-Ledger-$self->{login}=$self->{sessionid}; expires=$today; path=/;\n| if $self->{login};
+      
   }
 
 }
 
- 
+
 sub redirect {
   my ($self, $msg) = @_;
 
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/Inifile.pm /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/Inifile.pm
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/Inifile.pm	2006-08-17 01:00:42.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/Inifile.pm	2006-08-25 21:36:59.000000000 +0200
@@ -32,7 +32,6 @@
 sub new {
   my ($type, $file) = @_;
 
-  warn "$type has no copy constructor! creating a new object." if ref($type);
   $type = ref($type) || $type;
   my $self = bless {}, $type;
   $self->add_file($file) if defined $file;
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/Menu.pm /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/Menu.pm
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/Menu.pm	2006-01-02 00:30:55.000000000 +0100
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/Menu.pm	2006-09-09 19:43:33.000000000 +0200
@@ -39,7 +39,7 @@
   my $target = ($self->{$item}{target}) ? $self->{$item}{target} : "";
 
   my $level = $form->escape($item);
-  my $str = qq|<a href=$module?path=$form->{path}&action=$action&level=$level&login=$form->{login}&timeout=$form->{timeout}&sessionid=$form->{sessionid}&js=$form->{js}|;
+  my $str = qq|<a href=$module?path=$form->{path}&action=$action&level=$level&login=$form->{login}&js=$form->{js}|;
 
   my @vars = qw(module action target href);
   
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/OE.pm /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/OE.pm
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/OE.pm	2006-08-25 19:32:58.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/OE.pm	2006-09-03 00:55:34.000000000 +0200
@@ -2185,7 +2185,7 @@
       $ordnumber ||= $form->update_defaults($myconfig, $numberfld, $dbh);
     
       $query = qq|INSERT INTO oe (ordnumber)
-		  VALUES ($uid)|;
+		  VALUES ('$uid')|;
       $dbh->do($query) || $form->dberror($query);
 
       $query = qq|SELECT id
diff -Nru /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/User.pm /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/User.pm
--- /tmp/anJCgeNvlb/sql-ledger-2.6.17/SL/User.pm	2006-06-07 21:19:44.000000000 +0200
+++ /tmp/TmYG7t9ZOc/sql-ledger-2.6.18/SL/User.pm	2006-09-10 00:50:23.000000000 +0200
@@ -104,10 +104,11 @@
 	return -1;
       }
     }
+   
+    $self->{password} = $form->{password};
+    $self->create_config("$userspath/$self->{login}.conf");
     
-    unless (-f "$userspath/$self->{login}.conf") {
-      $self->create_config("$userspath/$self->{login}.conf");
-    }
+    $self->{password} = $form->{password};
     
     do "$userspath/$self->{login}.conf";
     $myconfig{dbpasswd} = unpack 'u', $myconfig{dbpasswd};
@@ -744,9 +745,37 @@
 sub create_config {
   my ($self, $filename) = @_;
 
-
   @config = &config_vars;
 
+  my $password = $self->{password};
+  my $key = "";
+  
+  $self->{sessionkey} = "";
+  $self->{sessioncookie} = "";
+  
+  if ($self->{password}) {
+    my $t = time + $self->{timeout};
+    srand( time() ^ ($$ + ($$ << 15)) );
+    $key = "$self->{login}$self->{password}$t";
+
+    my $i = 0;
+    my $l = length $key;
+    my $j = $l;
+    my %ndx = ();
+    my $pos;
+
+    while ($j > 0) {
+      $pos = int rand($l);
+      next if $ndx{$pos};
+      $ndx{$pos} = 1;
+      $self->{sessioncookie} .= substr($key, $pos, 1);
+      $self->{sessionkey} .= substr("0$pos", -2);
+      $j--;
+    }
+    
+    $self->{password} = crypt $self->{password}, substr($self->{login}, 0, 2) if ! $self->{encrypted};
+  }
+
   open(CONF, ">$filename") or $self->error("$filename : $!");
   
   # create the config file
@@ -766,6 +795,13 @@
 
   close CONF;
 
+  foreach $key (sort @config) {
+    $self->{$key} =~ s/\\\\/\\/g;
+    $self->{$key} =~ s/\\'/'/g;
+  }
+
+  $self->{password} = $password;
+  
 }
 
 
@@ -813,7 +849,8 @@
     chop $self->{dbpasswd};
   }
   
-  if ($self->{password} ne $self->{old_password}) {
+  my $password = $self->{password};
+  if (!$self->{encrypted}) {
     $self->{password} = crypt $self->{password}, substr($self->{login}, 0, 2) if $self->{password};
   }
   
@@ -821,6 +858,7 @@
     @config = qw(password);
   } else {
     @config = &config_vars;
+    @config = grep !/^session/, @config;
   }
  
   # replace \r\n with \n
@@ -834,10 +872,9 @@
   
   # create conf file
   if (! $self->{'root login'}) {
+    $self->{password} = $password;
     $self->create_config("$userspath/$self->{login}.conf");
 
-    $self->{dbpasswd} =~ s/\\'/'/g;
-    $self->{dbpasswd} =~ s/\\\\/\\/g;
     $self->{dbpasswd} = unpack 'u', $self->{dbpasswd};
     
     # check if login is in database
@@ -905,11 +942,11 @@
 
 
 sub config_vars {
-  
-  my @conf = qw(acs address businessnumber company countrycode
-             currency dateformat dbconnect dbdriver dbhost dbname dboptions
+ 
+  my @conf = qw(acs address businessnumber company countrycode currency
+             dateformat dbconnect dbdriver dbhost dbname dboptions
 	     dbpasswd dbport dbuser email fax menuwidth name numberformat
-	     password printer role sid signature stylesheet tel
+	     password printer sessionkey role sid signature stylesheet tel
 	     templates timeout vclimit);
 
   @conf;
